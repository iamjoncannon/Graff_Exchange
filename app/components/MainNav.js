import React from 'react';
import { connect } from "react-redux";
import { logoUrl } from './utils'
import { Link } from 'react-router-dom'
import Socket from './Socket'
import { isDesktop } from './utils'
import { hydratePortfolioThunk } from "../../store/Portfolio/thunks_for_Portfolio.js"
import UserActions from "../../store/User/actions_for_User.js"

// since the main nav gets mounted 
// when the user logs in, and its mounted
// through the application workflow, it will
// be used to retrieve initial data and will
// host the socket

class MainNav extends React.Component {
 
  constructor(props){
      super(props)

  } 

  componentDidMount(){

    // this.props.hydratePortfolio(this.props.token)
  }

  render(){

    const { pathname } = this.props.location;

    return (

      <div className="main-nav">

          <Socket />

        { isDesktop && 
    
          <div className="desk-nav">
              <span>Gopher Exchange</span>
              {/* <span>Account</span> */}
              <span>{this.props.Name}</span>
              <span>Balance: ${Number(this.props.Balance).toFixed(2)}</span>
              <span onClick={this.props.LogOut}>Log Out</span>
          </div> 
        }

        <div className="all-screens">
          
          <Link to="/all/watchlist">

            <i className={`fas fa-home fa-${isDesktop ? "2" : "7"}x ${ pathname.includes("/all") ? "selected" : ""}`} />

            {isDesktop() && <span>Portfolio</span>}
          
          </Link>

          <img src={logoUrl} />
          
    
            <Link to={ Object.keys(this.props.portfolio).length > 0 ? "/indiv/perf" : "/all/watchlist"}>

              {isDesktop() && <span>Holdings</span>}   

              <i className={`fas fa-chart-bar fa-${isDesktop ? "2" : "7"}x ${ pathname.includes("/indiv") ? "selected" : ""}`} />
            
            </Link>
 

        </div>

      </div>

    );
  }

};

const mapStateToProps = ({ User_state, Portfolio_state }) => {
  
  return {
    Name: User_state.Name,
    Balance: User_state.balance,
    token: User_state.token,
    portfolio: Portfolio_state.portfolio
  };
};

const mapDispatchToProps = dispatch => ({
  LogOut : () => dispatch(UserActions.LogOut()),
  hydratePortfolio: (token) => dispatch(hydratePortfolioThunk(token)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MainNav);