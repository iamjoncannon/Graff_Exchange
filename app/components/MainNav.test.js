import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import MainNav from './MainNav'
import configureStore from 'redux-mock-store';
import thunk from "redux-thunk"
const mockStore = configureStore([thunk]);
import { Router } from "react-router-dom";
import { createMemoryHistory } from 'history'

const fake_store = {
    User_state : {
        Name: "Jay",
        balance: 0, 
        token: ''
    }, 
    Portfolio_state : {
        portfolio: {

        }
    }
}

const store_setup = (store) => {

    const history = createMemoryHistory()

    history.push("/")
    
    return renderer.create(
    <Provider store={store} >
        <Router history={history}>    
            <MainNav location={ { pathname: "/"}}/>
        </Router>
    </Provider>
    );
}

describe("MainNav", ()=>{

    let component 
    let connected_react_component
    let store 

    beforeEach(()=>{
        
        window.outerWidth = 1101

        store = mockStore( fake_store );

        component = store_setup(store)

        connected_react_component = component.root.children[0].children[0].children[0].children[0]._fiber.stateNode
    })

    it('1) renders successfully with data from redux store', ()=>{

        let tree = component.toJSON();
        
        expect(tree).toMatchSnapshot();
    })

    it("2) if desktop, renders 'desk-nav' div with four spans",()=>{

        const desk_nav = component.root.find(node => node.props.className === "desk-nav")

        expect(desk_nav).toBeTruthy()

        expect(desk_nav.children.length).toEqual(4)
    })

    it("3) if desktop, renders spans containing Portfolio and Holdings",()=>{

        const spans = component.root.findAllByType('span').map(node => node.children[0])

        expect(spans.includes("Portfolio")).toBeTruthy()
        expect(spans.includes("Holdings")).toBeTruthy()
    })
    
    it("4) renders Socket component",()=>{

        const socket = component.root.find(node => node.type.name === "Socket")
    
        expect(socket).toBeTruthy()
    })
    
    it("5) renders fa-home and fa-chart icons",()=>{

        const icons = component.root.findAllByType("i").map(node => node.props.className)

        expect(icons.length).toEqual(2)
    })

    it("6) renders two React Router Link components",()=>{

        const links = component.root.findAll(node => node.type.displayName === "Link")

        expect(links.length).toEqual(2)
    })

    it("7) Portfolio Link redirects to /all/watchlist",()=>{

        const links = component.root.findAll(node => node.type.displayName === "Link")

        expect(links[0].props.to).toEqual('/all/watchlist')
    })
   
    it("8) Holding Link redirects to '/all/watchlist'",()=>{
        const links = component.root.findAll(node => node.type.displayName === "Link")

        expect(links[1].props.to).toEqual('/all/watchlist')
    })

})