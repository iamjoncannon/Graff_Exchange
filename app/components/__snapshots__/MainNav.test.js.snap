// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainNav 1) renders successfully with data from redux store 1`] = `
<div
  className="main-nav"
>
  <div
    style={
      Object {
        "height": "0",
        "width": "0",
      }
    }
  />
  <div
    className="desk-nav"
  >
    <span>
      Graff Exchange
    </span>
    <span>
      Jay
    </span>
    <span>
      Balance: $
      0.00
    </span>
    <span
      onClick={[Function]}
    >
      Log Out
    </span>
  </div>
  <div
    className="all-screens"
  >
    <a
      href="/all/watchlist"
      onClick={[Function]}
    >
      <i
        className="fas fa-home fa-2x "
      />
      <span>
        Portfolio
      </span>
    </a>
    <img
      src="../assets/graphql_logo.png"
    />
    <a
      href="/all/watchlist"
      onClick={[Function]}
    >
      <span>
        Holdings
      </span>
      <i
        className="fas fa-chart-bar fa-2x "
      />
    </a>
  </div>
</div>
`;
